Add C#-style type accelerators and suffixes for ushort, uint, ulong, and short literals.

vexx32 opened the pull request:
## PR Summary

This has been split off of #7509 at request of @iSazonov as that PR has become overly complex, and this portion of the PR was already approved by Powershell Committee.

Adds the following numeric literal suffixes:

* 'u' suffix (uint32)
* 's' suffix (short)
* 'ul' suffix (ulong)
* 'us' suffix (ushort)

Adds the following type accelerators:

* `[short]`
* `[ushort]`
* `[uint]`
* `[ulong]`

And of course, adds tests for all the above.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

vexx32 committed:
Merge pull request #1 from PowerShell/master

Catch up to PowerShell/PowerShell/master

vexx32 committed:
Merge pull request #2 from PowerShell/master

Catch up to master

vexx32 committed:
Merge branch 'master' of git://github.com/PowerShell/PowerShell

vexx32 committed:
Add typesuffix traits to u and s

vexx32 committed:
Add typesuffix traits to u and s

vexx32 committed:
Add typesuffix traits to u and s

vexx32 committed:
Add type accelerators for uint, ulong, ushort, and short

vexx32 committed:
Add type accelerators for uint, ulong, ushort, and short

vexx32 committed:
Add type accelerators for uint, ulong, ushort, and short

vexx32 committed:
Add support for parsing uint, ulong, ushort, and short values

vexx32 committed:
Add support for parsing uint, ulong, ushort, and short values

vexx32 committed:
Add support for parsing uint, ulong, ushort, and short values

vexx32 committed:
Add support for parsing uint, ulong, ushort, and short values

vexx32 committed:
Add tests for parsing the new literals

vexx32 committed:
Add tests for parsing the new literals

vexx32 committed:
Add tests for parsing the new literals

vexx32 committed:
Add tests for parsing the new literals

vexx32 committed:
Add tests for new type accelerators

vexx32 committed:
Add tests for new type accelerators

vexx32 committed:
Add tests for new type accelerators

vexx32 committed:
Overlooked a suffix reference I need to change the type of.

vexx32 committed:
Overlooked a suffix reference I need to change the type of.

vexx32 committed:
Overlooked a suffix reference I need to change the type of.

vexx32 committed:
Remove weird copy paste error
How did that even happen, seriously...

vexx32 committed:
Remove weird copy paste error
How did that even happen, seriously...

vexx32 committed:
Remove weird copy paste error
How did that even happen, seriously...

vexx32 committed:
Remove weird copy paste error
How did that even happen, seriously...

vexx32 commented:
I don't intend to fix the code factor issues here. I'll address them during the parser refactor that will come after this.

vexx32 commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -3268,13 +3278,33 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
                             else
                             {
-                                result = d * multiplier;
+                                result = d;
```
Ah, I thought I got all those. Correcting...

(I was trying to see if I could do the multiplier before the conversions to avoid throwing too much here, but it doesn't work *because* Int64.MaxValue is too high to retain precision with Double values in the upper range there.)

vexx32 commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -3268,13 +3278,33 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
-                                result = ((long)Convert.ChangeType(d, typeof(long), CultureInfo.InvariantCulture)) * multiplier;
+                                result = ((long)Convert.ChangeType(d, typeof(long), CultureInfo.InvariantCulture) * multiplier);
+                            }
+                            else if (suffix == NumberSuffixFlags.Short)
```
Hmmm. Agreed!

vexx32 commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -3304,6 +3334,33 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
+                    {
+                        if (short.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out short s))
+                        {
+                            result = s * (short)multiplier;
```
At present, that will throw an error and is expected. 

With some of the upcoming changes (that you can see if you build the version in the original PR) that instead simply registers as an invalid numeric literal, as I am able to verify first that the multiplier does not overflow the intended value type bounds without introducing a morass of if/else.

iSazonov replied:
Closed under next PR.

vexx32 commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -3304,6 +3334,33 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
+                    }
+                    else if (suffix.HasFlag(NumberSuffixFlags.Unsigned) && ulong.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out ulong u))
+                    {
+                        u *= (ulong)multiplier;
```
I am also pondering that myself, but I've not seen a completely necessary need to change it. It does seem to simplify calculations against other signed types, however; unsigned types often need a cast when doing multiplication with signed types.

iSazonov replied:
Closed.

vexx32 commented on test/powershell/Language/Parser/Parser.Tests.ps1:
```diff
@@ -660,7 +660,13 @@ foo``u{2195}abc
-			@{ Script = "0x100000000"; Expected = [int64]0x100000000 }
+            @{ Script = "0x100000000"; Expected = [int64]0x100000000 }
+            #Tests for short notation
+            @{ Script = "10s"; Expected = "10" }
```
That is probably wise. I'll take some time over lunch to add tests for value types. :)

vexx32 commented on test/powershell/Language/Parser/Parser.Tests.ps1:
```diff
@@ -686,6 +692,23 @@ foo``u{2195}abc
+
+        $testInvalidNumerals = @(
+            @{ Script = "16p" }
+            @{ Script = "20ux" }
```
Sure, if we want to. `x` is neither a valid type suffix nor a multiplier in PS, though, so testing for it is more or less the same as testing `16p` at the moment. We could go through the alphabet if you want. :)

iSazonov replied:
No :-) We should cover _all_ code paths. Do we cover?

vexx32 commented on test/powershell/Language/Parser/Parser.Tests.ps1:
```diff
@@ -686,6 +692,23 @@ foo``u{2195}abc
+            @{ Script = "21ss" }
+            @{ Script = "100ll" }
+            @{ Script = "150su" }
+            @{ Script = "10ds" }
```
Ah, good catch. Absolutely.

vexx32 replied:
@iSazonov Also noting down now before I forget:
Presently the literal `0x10d` is *legal* in PowerShell, although `Decimal.TryParse()` does *not* allow the hexadecimal syntax and that parse fails. Its actual return type is `Int32`. 

I plan on fixing this as I go through the parser rewrite after this PR.

iSazonov replied:
0x10d - have we tests for it?

vexx32 replied:
Oh, lol. I am clearly not *quite* up to speed this morning. Of *course* 0x10d is legal. D is a hexadecimal character... and that's exactly why decimals will never result from a hex literal ending in `d`. I'm gonna make a comment in the code that reflects this, because it's not immediately obvious why Decimal doesn't even bother to check for hex literals.

vexx32 replied:
(and yes I believe there are tests for it, but I'll pop one in the type checks just to be sure.)

vexx32 replied:
Added all these. :)

vexx32 committed:
Address review comments
Partially replace if stack with switches
Add missing multiplier.

vexx32 committed:
Address review comments
Partially replace if stack with switches
Add missing multiplier.

vexx32 committed:
Address review comments
Partially replace if stack with switches
Add missing multiplier.

vexx32 committed:
Address review comments
Partially replace if stack with switches
Add missing multiplier.

vexx32 commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -3304,6 +3334,33 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
                         result = null;
                         return false;
                     }
+                    else if (suffix == NumberSuffixFlags.Short)
```
Again, agreed. Sorted out. :)

vexx32 committed:
Add More Tests for new literal types.

vexx32 committed:
Add More Tests for new literal types.

vexx32 committed:
Add More Tests for new literal types.

vexx32 committed:
Add More Tests for new literal types.

vexx32 committed:
Add comments for decimal weirdness & some tests

vexx32 committed:
Add comments for decimal weirdness & some tests

vexx32 committed:
Add comments for decimal weirdness & some tests

vexx32 committed:
Add comments for decimal weirdness & some tests

vexx32 committed:
Add test for `ud` suffix that should fail

vexx32 committed:
Add test for `ud` suffix that should fail

vexx32 committed:
Add test for `ud` suffix that should fail

vexx32 committed:
Add test for `ud` suffix that should fail

vexx32 committed:
Fix short parsing

Apparently int16 * int16 always results in int32.

Wat.

vexx32 committed:
Fix short parsing

Apparently int16 * int16 always results in int32.

Wat.

vexx32 committed:
Fix short parsing

Apparently int16 * int16 always results in int32.

Wat.

vexx32 committed:
Fix short parsing

Apparently int16 * int16 always results in int32.

Wat.

vexx32 committed:
Forgot the decimal case. Fixed that short parsing.

vexx32 committed:
Forgot the decimal case. Fixed that short parsing.

vexx32 committed:
Forgot the decimal case. Fixed that short parsing.

vexx32 commented:
OK, pretty sure I got that short parsing sorted. I can understand doing operations in the console types can be freely converted, but if you're gonna specify `s` and get an int32, there's not much of a point!

rjmholt commented:
@iSazonov I've added the review tag back, since @BrucePay added back in the other PR.

I think the scope of these changes are larger than the committee originally thought, and since they are pretty fundamental changes to some core functionality, they probably deserve some scrutiny and a holistic consideration.

iSazonov commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -3255,6 +3255,8 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, Number
                     NumberStyles style = NumberStyles.AllowLeadingSign | NumberStyles.AllowDecimalPoint |
                                          NumberStyles.AllowExponent;
+
+                    // Decimal parser does not accept hex literals, and 'd' is a valid hex character, so will never be read as Decimal literal
```
We could add sample.

vexx32 replied:
In what sense do you mean? Just an example?

iSazonov replied:
Yes, like `0x10d`.

vexx32 replied:
Will do! 😄

iSazonov replied:
Closed.

iSazonov commented on test/powershell/Language/Parser/Parser.Tests.ps1:
```diff
@@ -709,6 +711,26 @@ foo``u{2195}abc
+        )
+        It "<Script> should result in an [<Expected>] value" -TestCases $testLiteralTypes {
+            param($Script, $Expected)
+            ExecuteCommand $Script | Should -BeOfType $Expected
```
I think it should be added to tests where we check values.

vexx32 replied:
Hmm... Oh, you mean with those others there like the long checks...

I know what you mean, but after running into the quick with Int16 values I felt it perhaps might be a little more prudent to directly check types rather than checking the way some of the other tests 'check type'. It's hard to guarantee the value won't just be cast into the same form when it may not be quite correct.

iSazonov replied:
We can test type before any operations on value.

vexx32 replied:
I'm not following, I think. The existing test for values doesn't consider type except in a few cases, and even in those the type checks are implicit, not explicit, and casting may occur.

iSazonov replied:
Can we do:
```powershell
	Context "Numerical Notations Tests (starting at line 2374 to line 2452)" {
        $testData = @(
			#Test various numbers using the standard notation.
            @{ Script = "0"; Expected = "0" }
			@{ Script = "-2"; Expected = "-2"; ExpectedType = [int32] }
			@{ Script = "2"; Expected = "2"; ExpectedType = [int32] }
			@{ Script = $([int32]::MaxValue); Expected = $([int32]::MaxValue); ExpectedType = [int32] }
			@{ Script = $([int32]::MinValue); Expected = $([int32]::MinValue); ExpectedType = [int32] }

...

        It "<Script> should return <Expected>" -TestCases $testData {
            param ( $Script, $Expected, $ExpectedType )
            $res = ExecuteCommand $Script
            $res | Should -Be $Expected
             $res | Should -BeType $ExpectedType
       }

vexx32 replied:
Roll it all into one.... Hm. Yeap, should be able to. I'll check into it shortly.

iSazonov replied:
If we continue in follow PR to add optinizations in the code and fix a hex sign it is important check not only parsing each string here but also test a result type. I think this will help a lot in follow PR avoid regressions. 
If you discover that we check here not all numeric formats please add new lines in the test.

iSazonov commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -3294,15 +3329,46 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
+                                else
+                                {
+                                    result = (uint)u;
+                                }
```
Maybe more clear:
```c#
                               else if (suffix.HasFlag(NumberSuffixFlags.Long) || u <= uint.MaxValue)  
                                {  
                                     result = (uint)u;
                                 }  
                                else  
                                {  
                                    // ulong
                                    result = u;
                                 }  
```

vexx32 replied:
Oh, you mean have the ulong case come last and alter the penultimate else if... Hmm. Yeah, OK, I can see the merit in that. Will do.

iSazonov replied:
Closed.

vexx32 committed:
Add clarifying comment to uint parsing cases

vexx32 committed:
Add clarifying comment to uint parsing cases

vexx32 committed:
Add clarifying comment to uint parsing cases

vexx32 committed:
Add clarifying comment to uint parsing cases

vexx32 committed:
Reorder parsing cases for unsigned ints

vexx32 committed:
Reorder parsing cases for unsigned ints

vexx32 committed:
Reorder parsing cases for unsigned ints

vexx32 committed:
Reorder parsing cases for unsigned ints

vexx32 committed:
Revise parser notation tests

vexx32 committed:
Revise parser notation tests

vexx32 committed:
Revise parser notation tests

vexx32 committed:
Revise parser notation tests

vexx32 committed:
Fix Hexadecimal tests
Yes, of course they get read back into decimal in memory
Also fixed expectedtypes from reading in string exponent values.

vexx32 committed:
Fix Hexadecimal tests
Yes, of course they get read back into decimal in memory
Also fixed expectedtypes from reading in string exponent values.

vexx32 committed:
Fix Hexadecimal tests
Yes, of course they get read back into decimal in memory
Also fixed expectedtypes from reading in string exponent values.

vexx32 committed:
Fix typo in test

vexx32 committed:
Fix typo in test

vexx32 committed:
Fix typo in test

vexx32 committed:
Fix typo in test

vexx32 committed:
Pull double conversions into full switch
Way better than if / else if / else

vexx32 committed:
Pull double conversions into full switch
Way better than if / else if / else

vexx32 committed:
Pull double conversions into full switch
Way better than if / else if / else

vexx32 committed:
Remove unnecessary parentheses.

vexx32 committed:
Remove unnecessary parentheses.

vexx32 committed:
Remove unnecessary parentheses.

iSazonov commented on test/powershell/Language/Parser/Parser.Tests.ps1:
```diff
@@ -649,48 +649,72 @@ foo``u{2195}abc
-			@{ Script = "-2KB"; Expected = "-2048" }
+			@{ Script = "0kb"; Expected = "0"; ExpectedType = [int] }
+			@{ Script = "1kb"; Expected = "1024"; ExpectedType = [int] }
+			@{ Script = "-2KB"; Expected = "-2048"; ExpectedType = [int] }
```
We don't test all code paths with suffixes. Ex.: 9.12Kb, -2lkb. Could you please enhance tests?

I suggest remove lines starting with "#Tests for the K suffix for numbers.". Then add test code like we did for expected type tests.
```powershell
Context "Numerical Notations Tests (starting at line 2374 to line 2452)" {
    BeforeAll {
$suffixes = @(
     @{ Name = ""; Value = 1 }
    @{ Name = "kb"; Value = 1024 }
    @{ Name = "MB"; Value = 1024*1024 }
    @{ Name = "Gb"; Value = 1024*1024*1024 }
    @{ Name = "Pb"; Value = 1024*1024*1024*1024 }
)
}

It "<Script> should return <Expected>, with type <ExpectedType>" -TestCases $testData {  
     param ( $Script, $Expected, $ExpectedType )  

foreach ($suffix in $Suffixes) {
    $NumWithSuffix = $Script + $suffix.Name
    $ExpectedNum = $Expected * $suffix.Value
    ExecuteCommand $NumWithSuffix | Should -BeOfType $ExpectedType
    ExecuteCommand $NumWithSuffix | Should -Be $ExpectedNum
}
}

vexx32 replied:
It's a good idea to do that, but since high values and high multipliers can result in overflow errors I think we're unfortunately a bit too likely to get failing tests by just iterating through the multipliers with everything (Pb overflows Int32, Gb overflows Int16, etc., etc.)

I've added a bunch, let me know if you think this is sufficient. 😄

iSazonov commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -3284,37 +3284,41 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, Number
                                     break;
-                                default:
-                                    if (suffix.HasFlag(NumberSuffixFlags.Unsigned))
+                                case NumberSuffixFlags.Unsigned | NumberSuffixFlags.Long:
```
| -> && ?

iSazonov commented on src/System.Management.Automation/engine/parser/tokenizer.cs:
```diff
@@ -3294,15 +3334,47 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
-                        }
-                        result = null;
-                        return false;
+                        case NumberSuffixFlags.Long:
```
What about "ul"?

vexx32 replied:
Presently, that one lives under the default case, inside the checks for the Unsigned flags.

iSazonov replied:
Sorry case is not HasFlag. Closed.

vexx32 committed:
Expand numeric parsing test cases

vexx32 committed:
Expand numeric parsing test cases

vexx32 committed:
Expand numeric parsing test cases

vexx32 committed:
Expand numeric parsing test cases

vexx32 committed:
Add explicit multiplier value tests

vexx32 committed:
Add explicit multiplier value tests

vexx32 committed:
Add explicit multiplier value tests

vexx32 committed:
Add explicit multiplier value tests

vexx32 committed:
Fix Tests' expected values.

vexx32 committed:
Fix Tests' expected values.

vexx32 committed:
Fix Tests' expected values.

vexx32 committed:
Fix Tests' expected values.

iSazonov commented:
@vexx32 Thanks for adding new test but they do not cover all code paths. Also we will add other code paths in the subsequent PR. That's why I suggested using a loop for suffixes.

vexx32 commented:
Hmm, I see your point there, but as mentioned that will cause its own issues. I'll comb through the code paths and put together some more thorough tests from the beginning, should be a bit simpler than trying to squint at all the tests and see what's missing. 😄

iSazonov commented:
@vexx32 We pass a string in `ExecuteCommand` and call AddScript there - I don't see side effects. Then we directly return a result object - again we shouldn't catch side effects.

vexx32 commented:
Sorry, poor choice of words @iSazonov.

I'm thinking on it... I need to expand the tests and organise them a bit regardless, but as long as I don't add cases for the `short` or `ushort` type-d literal tests we should be fine. (Multipliers above `kb` will fail if applied to `(u)short` values because they are outside the value range of those data types.)

iSazonov commented:
My understanding is that in ExecuteCommand a test string will be parsed. If we would wanted to test only parser we would not call ExecuteCommand at all. So we test not only parsing but conversion to value too.
As for multipliers above kb - if we expected a fail we should check it too. I think it is very useful if we plan add new custom ToInt()-like methods (without throw). It allow us to check all edge cases (overflows). We could add new parameter "ExpectedThrow" in the test.

vexx32 commented:
Hmm. Might be a bit janky determining programmatically which cases are meant to fail. I'll keep looking at it, but I think you might be on a good track there.

vexx32 committed:
Cover *all* the test cases!

vexx32 committed:
Cover *all* the test cases!

vexx32 committed:
Cover *all* the test cases!

vexx32 committed:
Cover *all* the test cases!

vexx32 commented:
@iSazonov So I did give what you recommended a shot, but I ran into a *lot* of issues trying to handle edge cases, where it should convert up to (u)long values, where it shouldn't convert up, where it should/shouldn't error.

I hope what I've got here is a happy medium between 'holy mother of test cases, Batman!' and 'what the hell is that weird loop doing'. 😉 

Let me know if you'd like me to expand it any; with all the edge cases on the numerals, handling it programmatically is liable to be just another painful piece of code to try to upkeep. At least with explicit declaration we know exactly what it's testing, why, and how. 😄

vexx32 committed:
Fix testcases reference
Teach me to try to restructure the test file too much! ;)

vexx32 committed:
Fix testcases reference
Teach me to try to restructure the test file too much! ;)

vexx32 committed:
Fix testcases reference
Teach me to try to restructure the test file too much! ;)

vexx32 committed:
Fix testcases reference
Teach me to try to restructure the test file too much! ;)

vexx32 committed:
Fix typos in tests

vexx32 committed:
Fix typos in tests

vexx32 committed:
Fix typos in tests

vexx32 committed:
Fix typos in tests

iSazonov reviewed (commented):
@vexx32 We also support '+' on first place in numerics. I think we should add such tests too.

vexx32 committed:
Add test cases for numbers beginning with +
Fix test cases for reals with certain suffixes

vexx32 committed:
Add test cases for numbers beginning with +
Fix test cases for reals with certain suffixes

vexx32 committed:
Add test cases for numbers beginning with +
Fix test cases for reals with certain suffixes

vexx32 committed:
Add test cases for numbers beginning with +
Fix test cases for reals with certain suffixes

vexx32 commented:
Added! 😄 

And fixed those real `0.0` cases that I apparently didn't notice.

iSazonov commented:
@rjmholt @SteveL-MSFT @daxian-dbw @BrucePay Please review the PR.

vexx32 committed:
Merge branch 'master' of git://github.com/PowerShell/PowerShell

vexx32 committed:
Merge branch 'master' of git://github.com/PowerShell/PowerShell

vexx32 committed:
Duplicate NamedAttributeArgumentAst class
as a framework for NamedMethodArgumentAst class

vexx32 committed:
Duplicate NamedAttributeArgumentAst class
as a framework for NamedMethodArgumentAst class

vexx32 committed:
Build framework for accepting and recognising named arguments for methods

vexx32 committed:
Build framework for accepting and recognising named arguments for methods

SteveL-MSFT commented:
@PowerShell/powershell-committee re-reviewed this and maintains the same recommendation to move forward with this PR

vexx32 committed:
Merge branch 'master' of git://github.com/PowerShell/PowerShell

vexx32 commented:
Hmm, haven't seen *that* error yet. Looks like an unrelated timeout issue, though. 😄

iSazonov commented:
@vexx32 Please rebase to fix CIs.

vexx32 committed:
Merge branch 'master' of git://github.com/PowerShell/PowerShell

vexx32 committed:
Merge branch 'Tokenizer/TypeSuffixes' of https://github.com/vexx32/PowerShell into Tokenizer/TypeSuffixes

vexx32 committed:
Revert changes to AST that crept in from another branch
No, really, I have no idea how the heck this happened.
It's gone now.

vexx32 committed:
Revert changes to AST that crept in from another branch
No, really, I have no idea how the heck this happened.
It's gone now.

vexx32 committed:
Manual revert ast.cs to original
Because I cannot find a distinct commit that actually modified this file

vexx32 committed:
Manual revert ast.cs to original
Because I cannot find a distinct commit that actually modified this file

vexx32 commented:
Attempting... this always seems far messier than it should be.

vexx32 committed:
Merge branch 'Tokenizer/TypeSuffixes' of https://github.com/vexx32/PowerShell into Tokenizer/TypeSuffixes

vexx32 commented:
Okay... @iSazonov apologies about the mess. Not sure what it is with me and rebasing, but I've yet to manage to figure it out and get it done cleanly. The Linux build seems to be failing on some kind of permissions issue with that build environment, not sure what that's about.

Everything else looks green, though. :)

iSazonov commented:
You should reset to your last commit and then rebase.
Or create new branch, copy your commits and open new PR.

vexx32 commented:
@iSazonov yeah, point taken. Rebasing is something I've yet to get to grips with. CLosing this for now, let's continue from a cleaner slate in #7813, where everything's all ready. :D
